core/src/services/ghaa/core.rs

```diff
@@ -0,0 +1,185 @@
+
+pub struct GhaaCore {
+    pub config: GhaaConfig,
+    pub client: HttpClient,
+}
+
+impl Debug for GhaaCore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("GhaaCore")
+            .field("owner", &self.config.owner)
+            .field("repo", &self.config.repo)
+            .finish_non_exhaustive()
+    }
+}
+impl GhaaCore {
+    pub async fn download_artifact(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let url: String = format!(
+            "{}/{}/{}/actions/artifacts/{}/zip",
+            DEFAULT_GHAA_ENDPOINT_SUFFIX, self.config.owner, self.config.repo, path
+        );
+
+        let mut req = Request::get(&url);
+
+        req = req.header(USER_AGENT, format!("opendal/{VERSION} (service ghaa)"));
+        req = req.header(HEADER_GITHUB_API_VERSION, "2022-11-28");
+        req = req.header(ACCEPT, "application/vnd.github.v3+json");
+
+        if let Some(auth) = &self.config.token {
+            req = req.header(AUTHORIZATION, format!("Bearer {}", auth))
+        }
+
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.client.send(req).await
+    }
+
+    pub async fn stat_artifact(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let url: String = format!(
+            "{}/{}/{}/actions/artifacts/{}",
+            DEFAULT_GHAA_ENDPOINT_SUFFIX, self.config.owner, self.config.repo, path
+        );
+
+        let mut req = Request::get(&url);
+        req = req.header(USER_AGENT, format!("opendal/{VERSION} (service ghaa)"));
+        req = req.header(HEADER_GITHUB_API_VERSION, "2022-11-28");
+        req = req.header(ACCEPT, "application/vnd.github.v3+json");
+
+        if let Some(auth) = &self.config.token {
+            req = req.header(AUTHORIZATION, format!("Bearer {}", auth))
+        }
+
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.client.send(req).await
+    }
+
+    pub async fn delete_artifact(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let url: String = format!(
+            "{}/{}/{}/actions/artifacts/{}",
+            DEFAULT_GHAA_ENDPOINT_SUFFIX, self.config.owner, self.config.repo, path
+        );
+
+        let mut req = Request::delete(&url);
+        req = req.header(USER_AGENT, format!("opendal/{VERSION} (service ghaa)"));
+        req = req.header(HEADER_GITHUB_API_VERSION, "2022-11-28");
+        req = req.header(ACCEPT, "application/vnd.github.v3+json");
+
+        if let Some(auth) = &self.config.token {
+            req = req.header(AUTHORIZATION, format!("Bearer {}", auth))
+        }
+
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.client.send(req).await
+    }
+
+    pub async fn list_artifacts_under_repo(
```

Comment:
I'm not sure about this. OpenDAL requires the services to have path-based name lookup. Maybe we don't need to support this?